app/src/main/java/org/openobservatory/ooniprobe/activity/RunTestActivity.java

```diff
@@ -0,0 +1,236 @@
+
+import java.util.ArrayList;
+import java.util.Observable;
+import java.util.Observer;
+import java.util.Set;
+
+import uk.co.chrisjenx.calligraphy.CalligraphyConfig;
+import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;
+
+public class RunTestActivity extends AppCompatActivity implements Observer {
+    private RecyclerView testUrlList;
+    private UrlListAdapter mUrlListAdapter;
+    private static ImageView testImage;
+    private static AppCompatButton runButton;
+    private static TextView title, testTitle, urls;
+    private ProgressBar test_progress;
+    private static String test_name;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_run_test);
+
+        // Get the intent that started this activity
+        Intent intent = getIntent();
+        Uri data = intent.getData();
+        if (Intent.ACTION_VIEW.equals(intent.getAction())) {
+            Uri uri = intent.getData();
+            String mv = uri.getQueryParameter("mv");
+            if (mv != null){
+                if (versionCompare(BuildConfig.VERSION_NAME, mv) >= 0) {
+                    Set<String> parameters = uri.getQueryParameterNames();
+                    String tn = uri.getQueryParameter("tn");
+                    String ta = uri.getQueryParameter("ta");
+                    String td = uri.getQueryParameter("td");
+                    String test = NetworkMeasurement.getTestName(this, tn);
+                    if (test.length() > 0){
+                        test_name = tn;
+                        configureScreen(td, ta);
+                    }
+                    else {
+                        Alert.alertDialog(this, getString(R.string.invalid_parameter), getString(R.string.test_name) +  " : " + tn);
+                    }
+                }
+                else {
+                    Alert.alertDialogTwoButtons(this, getString(R.string.ooniprobe_outdate), getString(R.string.ooniprobe_outdate_msg), new DialogInterface.OnClickListener() {
+                        public void onClick(DialogInterface dialog, int id) {
+                            final String appPackageName = getPackageName(); // getPackageName() from Context or Activity object
+                            try {
+                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
+                            } catch (android.content.ActivityNotFoundException anfe) {
+                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
+                            }
+                        }
+                    });
+                }
+            }
+        }
+    }
+
+    public void configureScreen(String td, String ta){
+        TestData.getInstance(this, this).addObserver(this);
```

Comment:
My understanding:

- when the app is started by the main activity, the `context` of `TestData` is the `MainActivity` and the `activity` of `TestData` is the main activity;

- when the app is started by this activity, `context` and `activity` are this activity.

I still haven't understood what happens when this activity exits. In the case in which the app also quits, then probably okay. Whereas, if the main activity is somehow started, wouldn't this mean that we're leaking this activity because it will never be freed since there is a reference loop between `TestData` and this activity and, moreover, `TestData` is referenced by a live object, i.e. the `MainActivity`?

Reply:
^^^ @lorenzoPrimi?

Reply:
Whenever you enter MainActivity I re-set the activity to that. Either MainActivity or RunTestActivity will be running

Reply:
> Either MainActivity or RunTestActivity will be running

Okay. And what happens when you quit `RunTestActivity`?

BTW: long time ago I remember discussing with @antoniolangiu that Android apps should use intents to communicate between backend services and user-facing activities. That was something that @antoniolangiu knew and it was not clear to me where this knowledge come from. Maybe, @antoniolangiu, can you share with us a couple of URLs related to that, so we can read a little bit, and possibly planning a refactoring oriented to using intents?